File: loggerRate.java
// TC: O(1)
// SC: O(N) store values in hashMap


import java.util.*;

public class loggerRate {

	HashMap<String, Integer> map = new HashMap<>();
;
	public boolean shouldPrintMessage(int time, String message) {
		
		if(!map.containsKey(message) || (time - map.get(message) >= 10) ) {
			map.put(message, time);
			return true;
		}
		return false;
	}
	
	public static void main(String[] args) {
		loggerRate logger = new loggerRate();
		System.out.println(logger.shouldPrintMessage(1, "foo")); //returns true; 

		// logging string "bar" at timestamp 2
		System.out.println(logger.shouldPrintMessage(2,"bar"));// returns true;

		// logging string "foo" at timestamp 3
		System.out.println(logger.shouldPrintMessage(3,"foo")); //returns false;

		// logging string "bar" at timestamp 8
		System.out.println(logger.shouldPrintMessage(8,"bar")); //returns false;

		// logging string "foo" at timestamp 10
		System.out.println(logger.shouldPrintMessage(10,"foo"));// returns false;

		// logging string "foo" at timestamp 11
		System.out.println(logger.shouldPrintMessage(11,"foo")); //returns true;
	}
}

File: loggerRate.py
'''
-- Passed test cases on LeetCode
'''
class Logger:
    #O(1) - Time complexity as lookup and update in hashmap takes constant time
    #O(n) - space complexity as in n represents the number of messages which are to be stored in the hashmap
    def __init__(self):
        self.dict = {}
        

    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        if message not in self.dict:
            self.dict[message] = timestamp
            return True
        if timestamp - self.dict[message] >= 10:
            self.dict[message] = timestamp
            return True
        return False
        


# Your Logger object will be instantiated and called as such:
# obj = Logger()
# param_1 = obj.shouldPrintMessage(timestamp,message)
